File: xander/MTAjs/mta.js
var lines = {
  'N': ["8th", "Union Square", "23rd", "28th", "34th", "Times Square"],
  'L': ["1st", "3rd", "Union Square", "6th", "8th"],
  '6': ["Astor Place", "Union Square", "23rd", "28th", "33rd", "Grand Central"]
};

var userInputCurrent = prompt('Where are you?');
var userInputDestination = prompt('Where are you going?');
var currentLine = userInputCurrent.slice(0,1);
var currentStation = userInputCurrent.slice(2);
var destinationLine = userInputDestination.slice(0,1);
var destinationStation = userInputDestination.slice(2);
var currentStartIndex;
var currentUnionSquare;
var destinationUnionSquare;
var destinationFinishIndex;
var stationsPassed;
var currentStationsPassed;
var destinationStationsPassed;
var changeLines;

if ( currentLine != 'L' && currentLine != 'N' && currentLine != '6') {
  console.log("That line doesn't exist! Please try again.");
  if (lines[currentLine].indexOf(currentStation) === -1 || lines[destinationLine].indexOf(destinationStation) === -1) {
      console.log("That station doesn't exist! Please try again.");
  }   
}

if (currentLine === destinationLine) {
  var startIndex = lines[currentLine].indexOf(currentStation);
  var finishIndex = lines[destinationLine].indexOf(destinationStation);
  changeLines = 'You will not be required to change lines on this journey.';
  
  if (startIndex < finishIndex) {
  stationsPassed = lines[currentLine].slice(startIndex, finishIndex); 
  } else {
    stationsPassed = lines[currentLine].slice(finishIndex, startIndex);  
  }
}

changeLines = 'You will need to change to line ' + destinationLine + ' at Union Square.';
currentStartIndex = lines[currentLine].indexOf(currentStation);
currentUnionSquare = lines[currentLine].indexOf("Union Square");
destinationUnionSquare = lines[destinationLine].indexOf("Union Square");
destinationFinishIndex = lines[destinationLine].indexOf(destinationStation); 
      
if (currentStartIndex < currentUnionSquare) {
  currentStationsPassed = lines[currentLine].slice(currentStartIndex, currentUnionSquare);
} else {
  currentStationsPassed = lines[currentLine].slice(currentUnionSquare, currentStartIndex);
  currentStationsPassed = currentStationsPassed.reverse();
}

if (destinationFinishIndex < destinationUnionSquare) {
  destinationStationsPassed = lines[destinationLine].slice(destinationFinishIndex, (destinationUnionSquare - 1));
  destinationStationsPassed = destinationStationsPassed.reverse();
} else {
  destinationStationsPassed = lines[destinationLine].slice((destinationUnionSquare + 1), destinationFinishIndex);
}
stationsPassed = currentStationsPassed.concat(destinationStationsPassed);
  
var numberOfStops = (stationsPassed.length);

console.log('You are travelling from ' + currentStation + ' station on Line ' + currentLine + ', to ' + destinationStation + ' station on Line ' + destinationLine + '. ' + changeLines + ' You will pass ' + numberOfStops + ' stations before you arrive at ' + destinationStation + ' station on Line ' + destinationLine + '.');

console.log('Your journey will encompass the following stops: ' + stationsPassed.join(', ') + '.')

File: tom/MTA/mta.rb
# currently only works with the N line hardcoded
# accepts current station and destination inputs
# doesn't seem to like strings with spaces in them
# goes backwards successfully

# subway stations are stored in a hash
subway = {
	'N' => ['Times Square', '34th', '28th', '23rd', 'Union Square', '3rd', '1st'],

	'L' => ['8th', '6th', 'Union Square', '3rd', '1st'],

	'6' => ['Grand Central', '33rd', '28th', '23rd', 'Union Square', 'Astor Place']
}

# save this for when I figure out how to use other lines as hash keys
# print "Please enter the line you are on: "
# user_line = gets.chomp.capitalize

# finds the index of user_station
print 'Please enter your current station: '
user_station = gets.chomp

start = subway['N'].index {|x| x == user_station}

# finds the index of user_destination
print 'Please enter your destination: '
user_destination = gets.chomp.capitalize

finish = subway['N'].index {|x| x == user_destination}

# next we need to compare the two indexes
if start < finish
	stop_count = finish - start
else
	stop_count = start - finish
end

puts "You need to stay on for #{stop_count} stops."

# loops through every value/index pair in the array of stations for the line in question
# checks each index to see if it is within the range of start to finish (variables containing the relevant indexes)
reverse_array =[]
# prints out relevant variables
print "You will pass "
subway['N'].each_with_index do |v, i| 
	if start < finish 
		unless i <= start || i > finish
			if i == finish
				print 'and end your journey at ' + v + '.'
				print "\n"
			else
				print v + ', '
			end		
		end
	# when finish > start
	else
		# puts the relevant items in an array
		unless i >= start || i < finish
			reverse_array << v
		end
	end
end

# to avoid messing with an already working loop above:
# - if statement checks to see if we're going backwards
# - reverses the stops
# - prints out the reversed stops
if finish < start
	reverse_array.reverse!
	
	reverse_array.each_with_index do |v, i|
		if i != reverse_array.index(reverse_array.last)
			print v + ', '
		else
			print 'and end your journey at ' + v + '.'
			print "\n"
		end

	end
end

# next step is to figure out how to input a line to travel on